[debug] Detailed reset reason

edwardtfn opened the pull request:
# What does this implement/fix?

This implements the use of `esp_reset_reason()` when `rtc_get_reset_reason()` returned "Software reset CPU", which is the huge majority of the cases.
With that, users will have additional info about the reset reasons.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):** N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** N/A

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

N/A
It will be available when using regular `debug` component.

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

edwardtfn committed:
Use `esp_reset_reason()` for `reset_reason`

This provides more details than `rtc_get_reset_reason(0)`, making the troubleshooting process a bit easier.

edwardtfn committed:
I forgot the `;`

No comments, please. :stuck_out_tongue_winking_eye:

edwardtfn committed:
Handle new entries when using ESP-IDF v4.4

edwardtfn committed:
Clean-up the strings

edwardtfn committed:
Lint

edwardtfn committed:
Adjust flags for arduino

edwardtfn committed:
clang

rvalitov commented on esphome/components/debug/debug_esp32.cpp:
```diff
@@ -85,7 +85,61 @@ std::string DebugComponent::get_reset_reason_() {
+        case ESP_RST_SDIO:
+          reset_reason = "Reset over SDIO";
+          break;
+#if defined(USE_ESP_IDF) && ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 1, 0)
```
```suggestion
#ifdef USE_ESP32
#if (ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 4, 5))
```
- The preprocessor checks should be divided into multiple lines to work properly.
- Is there any specific need setting minimum version to 5.1? I see the used constants are available since version [4.2](https://github.com/espressif/esp-idf/blob/release/v4.2/components/esp_system/include/esp_system.h). However, in my tests it's actually compiles and works since version 4.4.5 that contains all necessary bug fixes. So I suggest putting 4.4.5 here.

edwardtfn replied:
Not all the constants are available prior to v5.1. The ones above this #if are the ones available since earlier versions.

edwardtfn replied:
And to be honest, I think this should be reversed in a way this `esp_reset_reason` first and only when that is unknown to use the other method.
Any thoughts?

edwardtfn replied:
I've reverted this to v5.1.0 as those advanced options are not available before IDF v5.1.0 and therefore causes error when compiling with ESP-IDF default version (4.4.8).

rvalitov commented on esphome/components/debug/debug_esp32.cpp:
```diff
@@ -85,7 +85,61 @@ std::string DebugComponent::get_reset_reason_() {
+        case ESP_RST_CPU_LOCKUP:
+          reset_reason = "Reset due to CPU lock up (double exception)";
+          break;
+#endif
```
```suggestion
#endif  // ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 4, 5)
#endif  // USE_ESP32
```

edwardtfn committed:
Update esphome/components/debug/debug_esp32.cpp

Co-authored-by: Ramil Valitov <ramilvalitov@gmail.com>

edwardtfn committed:
Update esphome/components/debug/debug_esp32.cpp

Co-authored-by: Ramil Valitov <ramilvalitov@gmail.com>

edwardtfn commented:
I think this should be reversed in a way this `esp_reset_reason` first and only when that is unknown to use the other (original) method.
Any thoughts?

edwardtfn committed:
Use USE_ESP32_VARIANT_ESP32 only for newer keys

edwardtfn committed:
Inverted methods

edwardtfn committed:
clang-format

edwardtfn commented:
I'm not an expert on libretiny... why is it trying to use this code? It should be only for ESP32.

rvalitov commented on esphome/components/debug/debug_esp32.cpp:
```diff
@@ -30,109 +30,162 @@ static const char *const TAG = "debug";
+    case ESP_RST_SDIO:
+      reset_reason = "Reset over SDIO";
+      break;
+#ifdef USE_ESP32_VARIANT_ESP32
```
```suggestion
#ifdef USE_ESP32
```
Required to pass the build checks

rvalitov commented on esphome/components/debug/debug_esp32.cpp:
```diff
@@ -30,109 +30,162 @@ static const char *const TAG = "debug";
+      reset_reason = "Reset over SDIO";
+      break;
+#ifdef USE_ESP32_VARIANT_ESP32
+#if (ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 4, 5))
```
```suggestion
#if (ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 1, 0))
```
Yes, you're right, 5.1.0 is required here.

rvalitov commented on esphome/components/debug/debug_esp32.cpp:
```diff
@@ -30,109 +30,162 @@ static const char *const TAG = "debug";
+    case ESP_RST_CPU_LOCKUP:
+      reset_reason = "Reset due to CPU lock up (double exception)";
+      break;
+#endif        // ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 4, 5)
```
```suggestion
#endif        // ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 1, 0)
```

rvalitov commented on esphome/components/debug/debug_esp32.cpp:
```diff
@@ -30,109 +30,162 @@ static const char *const TAG = "debug";
+      reset_reason = "Reset due to CPU lock up (double exception)";
+      break;
+#endif        // ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 4, 5)
+#endif        // USE_ESP32_VARIANT_ESP32
```
```suggestion
#endif        // USE_ESP32
```

edwardtfn committed:
Advanced options only on IDF 5.1+

rvalitov commented:
I tested locally with the following config:

```yaml
esphome:
  name: reset-reason-test-arduino
  friendly_name: Test

bk72xx:
  board: generic-bk7231n-qfn32-tuya

external_components:
  - source: github://pr#7729
    components: [ debug ]

# Enable logging
logger:
  level: VERBOSE

debug:
  update_interval: 30s

text_sensor:
  - platform: debug
    device:
      name: "Device Info"
    reset_reason:
      name: "Reset Reason"
```

It builds successfully:
```console
INFO ESPHome 2024.11.0-dev
INFO Reading configuration reset-reason-arduino.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing reset-reason-test-arduino (board: generic-bk7231n-qfn32-tuya; framework: arduino; platform: libretiny @ 1.7.0)
...

```

jesserockz commented:
> I'm not an expert on libretiny... why is it trying to use this code? It should be only for ESP32.

As per the failed CI, there are some reasons added here that are probably not available to all esp32 variants. libretiny has nothing to do with it. 

```
src/esphome/components/debug/debug_esp32.cpp:66:10: error: 'ESP_RST_USB' was not declared in this scope; did you mean 'ESP_RST_SW'?
   66 |     case ESP_RST_USB:
      |          ^~~~~~~~~~~
      |          ESP_RST_SW
src/esphome/components/debug/debug_esp32.cpp:69:10: error: 'ESP_RST_JTAG' was not declared in this scope; did you mean 'ESP_RST_WDT'?
   69 |     case ESP_RST_JTAG:
      |          ^~~~~~~~~~~~
      |          ESP_RST_WDT
src/esphome/components/debug/debug_esp32.cpp:72:10: error: 'ESP_RST_EFUSE' was not declared in this scope; did you mean 'ESP_RST_SW'?
   72 |     case ESP_RST_EFUSE:
      |          ^~~~~~~~~~~~~
      |          ESP_RST_SW
src/esphome/components/debug/debug_esp32.cpp:75:10: error: 'ESP_RST_PWR_GLITCH' was not declared in this scope
   75 |     case ESP_RST_PWR_GLITCH:
      |          ^~~~~~~~~~~~~~~~~~
src/esphome/components/debug/debug_esp32.cpp:78:10: error: 'ESP_RST_CPU_LOCKUP' was not declared in this scope
   78 |     case ESP_RST_CPU_LOCKUP:
      |          ^~~~~~~~~~~~~~~~~~
```

edwardtfn committed:
Advanced consts available from v5.1.4

https://github.com/espressif/esp-idf/blob/d7b0a45ddbddbac53afb4fc28168f9f9259dbb79/components/esp_system/include/esp_system.h#L36

edwardtfn committed:
Merge branch 'dev' into debug-reset-reason-detailed

edwardtfn commented:
Fixed! 😃

edwardtfn committed:
Merge branch 'dev' into debug-reset-reason-detailed
